Sources/PackagePlugin/Implementation.swift

```diff
@@ -0,0 +1,132 @@
+
+extension Plugin {
+
+    public static func main() throws {
+        // Look for the input JSON as the last argument of the invocation.
+        guard let inputData = ProcessInfo.processInfo.arguments.last?.data(using: .utf8) else {
+            fputs("Expected last argument to contain JSON input data in UTF-8 encoding, but didn't find it.", stderr)
+            output.diagnostics.append(Diagnostic(severity: .error, message: "Expected last argument to contain JSON input data in UTF-8 encoding, but didn't find it.", file: nil, line: nil))
+            exit(1)
+        }
+ 
+        // Decode the input JSON into a plugin context.
+        let context: TargetBuildContext
+        do {
+            let decoder = JSONDecoder()
+            context = try decoder.decode(TargetBuildContext.self, from: inputData)
+        }
+        catch {
+            fputs("Couldn't decode input JSON (reason: \(error)", stderr)
+            output.diagnostics.append(Diagnostic(severity: .error, message: "\(error)", file: nil, line: nil))
+            exit(1)
+        }
+
+        // Instantiate the plugin. For now there are no parameters, but this is where we would set them up, most likely as properties of the plugin.
+        let plugin = self.init()
+        
+        // Invoke the appropriate protocol method based on the action.
+        switch context.pluginAction {
+        case .createBuildToolCommands:
+            // Check that the plugin conforms to `BuildToolPlugin`, and get the commands.
+            guard let plugin = plugin as? BuildToolPlugin else { throw PluginDeserializationError.malformedInputJSON("...") }
+            let commands = try plugin.createBuildCommands(context: context)
+            
+            // Convert the commands to the encodable output representation SwiftPM currently expects.
+            output.buildCommands = commands.compactMap {
+                guard case let ._buildCommand(displayName, executable, arguments, environment, workingDir, inputFiles, outputFiles) = $0 else { return .none }
+                return BuildCommand(displayName: displayName, executable: executable, arguments: arguments, environment: environment, workingDirectory: workingDir, inputFiles: inputFiles, outputFiles: outputFiles)
+            }
+            output.prebuildCommands = commands.compactMap {
+                guard case let ._prebuildCommand(displayName, executable, arguments, environment, workingDir, outputFilesDir) = $0 else { return .none }
+                return PrebuildCommand(displayName: displayName, executable: executable, arguments: arguments, environment: environment, workingDirectory: workingDir, outputFilesDirectory: outputFilesDir)
+            }
+        }
+
+        // Emit the output from the plugin for SwiftPM to read.
+        let encoder = JSONEncoder()
+        let outputData = try! encoder.encode(output)
+        fputc(0, stdout)
+        fputs(String(data: outputData, encoding: .utf8)!, stdout)
+        fflush(stdout)
```

Comment:
are the printing functions here supported cross platform? is there a foundation version we can use?

Reply:
First, these aren't new to this diff but just moved.  Second, I'll take a look — we can probably use some Foundation function instead.  It's a good point to not convert to a string if we already have data, and we should certainly avoid a force unwrap.

Reply:
This is a bit subtle, and I remember now why I originally used `fputs()`.  The issue is that `print()` uses buffered stdout, and so we still do need to flush output here to make sure that the null and the JSON follow any output from the process (which is the point).  Using Foundation for this is a bit ugly, as flushing output can also throw in the general case (in practice not with stdout).

To answer your original question, yes this is supported.  @compnerd Are there any problems using `fwrite()` on Windows?  Would there be a benefit to using Foundation here?

Reply:
Actually this gets worse.  `try FileHandle.standardOutput.synchronize()` raises with file handles associated with pipes.  So we still need the `fflush()`.

This has been working fine since forever with package manifests, so I'm inclined to stay with what we have.  I can open a ticket on Foundation to support `synchronize` for pipes such as `stdout` as well.

But I will eliminate the redundant UTF-8 conversion.